File: frontend/src/lib/rbac.ts
import type { Role, AuthUser } from "@/stores/auth";

// ─────────────────────────────────────────────────────────────────────────────
// CANONICAL PERMISSIONS — kept in sync with backend app/core/rbac.py
// Do NOT add role-specific logic here. One source of truth only.
// ─────────────────────────────────────────────────────────────────────────────
export const PERMISSIONS = {
  // Operational / investigative
  DASHBOARD: "DASHBOARD",
  SEARCH_CASES: "SEARCH_CASES",
  CHAT_ASSISTANT: "CHAT_ASSISTANT",
  CRIME_MAP: "CRIME_MAP",
  PATTERN_INTELLIGENCE: "PATTERN_INTELLIGENCE",
  CRIMINAL_NETWORK: "CRIMINAL_NETWORK",
  FINANCIAL_CRIME: "FINANCIAL_CRIME",
  GANG_DETECTION: "GANG_DETECTION",
  SENSITIVE_CASE_ACCESS: "SENSITIVE_CASE_ACCESS",
  EXPORT_REPORTS: "EXPORT_REPORTS",

  // Supervisory
  VIEW_AUDIT_LOGS: "VIEW_AUDIT_LOGS",
  ASSIGN_DISTRICTS: "ASSIGN_DISTRICTS",
  APPROVE_ACCESS_REQUESTS: "APPROVE_ACCESS_REQUESTS",

  // Administrative
  MANAGE_USERS: "MANAGE_USERS",
  MANAGE_ROLES: "MANAGE_ROLES",
  SYSTEM_CONFIGURATION: "SYSTEM_CONFIGURATION",

  // ── Legacy aliases kept for backward compatibility with existing components ──
  // Do NOT use these in new code. Use the canonical names above.
  /** @deprecated use PATTERN_INTELLIGENCE */
  PATTERN_ANALYSIS: "PATTERN_INTELLIGENCE",
  /** @deprecated use EXPORT_REPORTS */
  EXPORT_DATA: "EXPORT_REPORTS",
} as const;

export type Permission = (typeof PERMISSIONS)[keyof typeof PERMISSIONS];

export type RouteAccess = {
  path: string;
  moduleName: string;
  permissions?: Permission[];
  roles?: Role[];
  public?: boolean;
};

// ─────────────────────────────────────────────────────────────────────────────
// ROLE → PERMISSIONS MAP
// Exact mirror of backend ROLE_PERMISSIONS in app/core/rbac.py
// ADMINISTRATOR is a platform admin — NOT an investigator.
// ─────────────────────────────────────────────────────────────────────────────
export const ROLE_PERMISSIONS: Record<Role, Permission[]> = {
  ADMINISTRATOR: [
    PERMISSIONS.VIEW_AUDIT_LOGS,
    PERMISSIONS.MANAGE_USERS,
    PERMISSIONS.MANAGE_ROLES,
    PERMISSIONS.SYSTEM_CONFIGURATION,
  ],

  SUPERVISOR: [
    PERMISSIONS.DASHBOARD,
    PERMISSIONS.SEARCH_CASES,
    PERMISSIONS.CHAT_ASSISTANT,
    PERMISSIONS.CRIME_MAP,
    PERMISSIONS.PATTERN_INTELLIGENCE,
    PERMISSIONS.CRIMINAL_NETWORK,
    PERMISSIONS.FINANCIAL_CRIME,
    PERMISSIONS.GANG_DETECTION,
    PERMISSIONS.SENSITIVE_CASE_ACCESS,
    PERMISSIONS.VIEW_AUDIT_LOGS,
    PERMISSIONS.EXPORT_REPORTS,
    PERMISSIONS.ASSIGN_DISTRICTS,
    PERMISSIONS.APPROVE_ACCESS_REQUESTS,
  ],

  SENIOR_INVESTIGATOR: [
    PERMISSIONS.DASHBOARD,
    PERMISSIONS.SEARCH_CASES,
    PERMISSIONS.CHAT_ASSISTANT,
    PERMISSIONS.CRIME_MAP,
    PERMISSIONS.PATTERN_INTELLIGENCE,
    PERMISSIONS.CRIMINAL_NETWORK,
    PERMISSIONS.FINANCIAL_CRIME,
    PERMISSIONS.GANG_DETECTION,
    PERMISSIONS.SENSITIVE_CASE_ACCESS,
    PERMISSIONS.EXPORT_REPORTS,
  ],

  INVESTIGATOR: [
    PERMISSIONS.DASHBOARD,
    PERMISSIONS.SEARCH_CASES,
    PERMISSIONS.CHAT_ASSISTANT,
    PERMISSIONS.CRIME_MAP,
    PERMISSIONS.PATTERN_INTELLIGENCE,
    PERMISSIONS.CRIMINAL_NETWORK,
    PERMISSIONS.EXPORT_REPORTS,
  ],

  ANALYST: [
    PERMISSIONS.DASHBOARD,
    PERMISSIONS.SEARCH_CASES,
    PERMISSIONS.CHAT_ASSISTANT,
    PERMISSIONS.CRIME_MAP,
    PERMISSIONS.PATTERN_INTELLIGENCE,
    PERMISSIONS.EXPORT_REPORTS,
  ],

  POLICY_MAKER: [
    PERMISSIONS.DASHBOARD,
    PERMISSIONS.SEARCH_CASES,
    PERMISSIONS.CHAT_ASSISTANT,
    PERMISSIONS.CRIME_MAP,
    PERMISSIONS.PATTERN_INTELLIGENCE,
  ],
};

// ─────────────────────────────────────────────────────────────────────────────
// ROUTE ACCESS — maps paths to required permissions for route guards.
// The _app.tsx beforeLoad checks these to redirect unauthorized access.
// ─────────────────────────────────────────────────────────────────────────────
export const ROUTE_ACCESS: RouteAccess[] = [
  { path: "/", moduleName: "Chat Assistant", permissions: [PERMISSIONS.CHAT_ASSISTANT] },
  { path: "/dashboard", moduleName: "Dashboard", permissions: [PERMISSIONS.DASHBOARD] },
  { path: "/cases", moduleName: "Case Search", permissions: [PERMISSIONS.SEARCH_CASES] },
  { path: "/offenders", moduleName: "Offender Profiles", permissions: [PERMISSIONS.SEARCH_CASES] },
  { path: "/network", moduleName: "Criminal Network", permissions: [PERMISSIONS.CRIMINAL_NETWORK] },
  { path: "/financial", moduleName: "Financial Crime", permissions: [PERMISSIONS.FINANCIAL_CRIME] },
  { path: "/map", moduleName: "Crime Map", permissions: [PERMISSIONS.CRIME_MAP] },
  { path: "/sociological", moduleName: "Pattern Intelligence", permissions: [PERMISSIONS.PATTERN_INTELLIGENCE] },
  { path: "/alerts", moduleName: "Investigation Alerts", permissions: [PERMISSIONS.PATTERN_INTELLIGENCE] },
  { path: "/audit", moduleName: "Audit Log", permissions: [PERMISSIONS.VIEW_AUDIT_LOGS] },
  // Supervisor Panel requires supervisory permissions — NOT admin permissions
  {
    path: "/supervisor",
    moduleName: "Supervisor Panel",
    permissions: [PERMISSIONS.ASSIGN_DISTRICTS, PERMISSIONS.APPROVE_ACCESS_REQUESTS],
  },
  { path: "/admin", moduleName: "Admin Console", permissions: [PERMISSIONS.MANAGE_USERS] },
  { path: "/settings", moduleName: "Settings", public: true },
  { path: "/access-restricted", moduleName: "Access Restricted", public: true },
];

// ─────────────────────────────────────────────────────────────────────────────
// DEFAULT LANDING PAGE per role
// Used by _app.tsx to redirect users to their appropriate home page after login.
// ─────────────────────────────────────────────────────────────────────────────
export const ROLE_DEFAULT_LANDING: Record<Role, string> = {
  ADMINISTRATOR: "/admin",
  SUPERVISOR: "/dashboard",
  SENIOR_INVESTIGATOR: "/dashboard",
  INVESTIGATOR: "/dashboard",
  ANALYST: "/dashboard",
  POLICY_MAKER: "/dashboard",
};

// ─────────────────────────────────────────────────────────────────────────────
// PERMISSION HELPERS
// ─────────────────────────────────────────────────────────────────────────────

export function getPermissionsForRole(role?: Role): Permission[] {
  return role ? (ROLE_PERMISSIONS[role] ?? []) : [];
}

export function hasPermission(
  user: Pick<AuthUser, "role" | "permissions"> | null | undefined,
  permission: Permission,
) {
  if (!user) return false;
  const effectivePerms = user.permissions?.length ? user.permissions : getPermissionsForRole(user.role);
  return effectivePerms.includes(permission);
}

export function hasAnyPermission(
  user: Pick<AuthUser, "role" | "permissions"> | null | undefined,
  permissions: Permission[] = [],
) {
  if (permissions.length === 0) return true;
  return permissions.some((permission) => hasPermission(user, permission));
}

export function hasAllPermissions(
  user: Pick<AuthUser, "role" | "permissions"> | null | undefined,
  permissions: Permission[] = [],
) {
  if (permissions.length === 0) return true;
  return permissions.every((permission) => hasPermission(user, permission));
}

export function canAccessRoute(user: AuthUser | null | undefined, route: RouteAccess) {
  if (route.public) return true;
  if (!user) return false;
  const roleOk = !route.roles?.length || route.roles.includes(user.role);
  return roleOk && hasAnyPermission(user, route.permissions);
}

export function findRouteAccess(pathname: string) {
  return ROUTE_ACCESS
    .filter((route) => pathname === route.path || pathname.startsWith(route.path + "/"))
    .sort((a, b) => b.path.length - a.path.length)[0];
}

// ─────────────────────────────────────────────────────────────────────────────
// HUMAN-READABLE LABELS
// ─────────────────────────────────────────────────────────────────────────────
const PERMISSION_LABELS: Partial<Record<Permission, string>> = {
  [PERMISSIONS.SEARCH_CASES]: "Investigator, Analyst, Supervisor, Policy Maker",
  [PERMISSIONS.CRIMINAL_NETWORK]: "Investigator, Senior Investigator, or Supervisor",
  [PERMISSIONS.CRIME_MAP]: "Any authenticated SCRB role",
  [PERMISSIONS.PATTERN_INTELLIGENCE]: "Investigator, Analyst, Supervisor, or Policy Maker",
  [PERMISSIONS.CHAT_ASSISTANT]: "Any authenticated SCRB role",
  [PERMISSIONS.FINANCIAL_CRIME]: "Senior Investigator or Supervisor",
  [PERMISSIONS.GANG_DETECTION]: "Senior Investigator or Supervisor",
  [PERMISSIONS.SENSITIVE_CASE_ACCESS]: "Senior Investigator or Supervisor",
  [PERMISSIONS.VIEW_AUDIT_LOGS]: "Supervisor or Administrator",
  [PERMISSIONS.MANAGE_USERS]: "Administrator",
  [PERMISSIONS.SYSTEM_CONFIGURATION]: "Administrator",
  [PERMISSIONS.EXPORT_REPORTS]: "Investigator, Senior Investigator, Supervisor, or Analyst",
  [PERMISSIONS.ASSIGN_DISTRICTS]: "Supervisor",
  [PERMISSIONS.APPROVE_ACCESS_REQUESTS]: "Supervisor",
  [PERMISSIONS.DASHBOARD]: "Investigator, Analyst, Supervisor, Senior Investigator, or Policy Maker",
};

export function requiredRoleLabel(permissions?: Permission[]) {
  if (!permissions?.length) return "Authorized SCRB personnel";
  return permissions
    .map((p) => PERMISSION_LABELS[p] ?? p)
    .join(" or ");
}

// ─────────────────────────────────────────────────────────────────────────────
// DATA MASKING UTILITIES
// ─────────────────────────────────────────────────────────────────────────────
export function maskName(value?: string | null) {
  if (!value) return "Hidden";
  const trimmed = value.trim();
  if (trimmed.length <= 2) return trimmed[0] + "*";
  return `${trimmed[0]}${"*".repeat(Math.min(trimmed.length - 2, 8))}${trimmed[trimmed.length - 1]}`;
}

export function maskPhone(value?: string | null) {
  if (!value) return "Hidden";
  const digits = value.replace(/\D/g, "");
  if (digits.length < 4) return "Hidden";
  return `${digits.slice(0, 2)}${"*".repeat(Math.max(2, digits.length - 4))}${digits.slice(-2)}`;
}

export function maskAccount(value?: string | null) {
  if (!value) return "Hidden";
  const clean = value.replace(/\s/g, "");
  return `XXXXXX${clean.slice(-4)}`;
}

File: backend/app/core/rbac.py
from enum import Enum


class UserRole(str, Enum):
    ADMINISTRATOR = "ADMINISTRATOR"
    SUPERVISOR = "SUPERVISOR"
    SENIOR_INVESTIGATOR = "SENIOR_INVESTIGATOR"
    INVESTIGATOR = "INVESTIGATOR"
    ANALYST = "ANALYST"
    POLICY_MAKER = "POLICY_MAKER"


class Permission:
    DASHBOARD = "DASHBOARD"
    SEARCH_CASES = "SEARCH_CASES"
    CHAT_ASSISTANT = "CHAT_ASSISTANT"
    CRIME_MAP = "CRIME_MAP"
    PATTERN_INTELLIGENCE = "PATTERN_INTELLIGENCE"
    CRIMINAL_NETWORK = "CRIMINAL_NETWORK"
    FINANCIAL_CRIME = "FINANCIAL_CRIME"
    GANG_DETECTION = "GANG_DETECTION"
    SENSITIVE_CASE_ACCESS = "SENSITIVE_CASE_ACCESS"
    VIEW_AUDIT_LOGS = "VIEW_AUDIT_LOGS"
    MANAGE_USERS = "MANAGE_USERS"
    MANAGE_ROLES = "MANAGE_ROLES"
    SYSTEM_CONFIGURATION = "SYSTEM_CONFIGURATION"
    EXPORT_REPORTS = "EXPORT_REPORTS"
    ASSIGN_DISTRICTS = "ASSIGN_DISTRICTS"
    APPROVE_ACCESS_REQUESTS = "APPROVE_ACCESS_REQUESTS"


# Centralized permission matrix enforcing the principal of least privilege
ROLE_PERMISSIONS = {
    UserRole.ADMINISTRATOR: {
        Permission.VIEW_AUDIT_LOGS,
        Permission.MANAGE_USERS,
        Permission.MANAGE_ROLES,
        Permission.SYSTEM_CONFIGURATION,
        Permission.ASSIGN_DISTRICTS,
    },
    UserRole.SUPERVISOR: {
        Permission.DASHBOARD,
        Permission.SEARCH_CASES,
        Permission.CHAT_ASSISTANT,
        Permission.CRIME_MAP,
        Permission.PATTERN_INTELLIGENCE,
        Permission.CRIMINAL_NETWORK,
        Permission.FINANCIAL_CRIME,
        Permission.GANG_DETECTION,
        Permission.SENSITIVE_CASE_ACCESS,
        Permission.VIEW_AUDIT_LOGS,
        Permission.EXPORT_REPORTS,
        Permission.ASSIGN_DISTRICTS,
        Permission.APPROVE_ACCESS_REQUESTS,
    },
    UserRole.SENIOR_INVESTIGATOR: {
        Permission.DASHBOARD,
        Permission.SEARCH_CASES,
        Permission.CHAT_ASSISTANT,
        Permission.CRIME_MAP,
        Permission.PATTERN_INTELLIGENCE,
        Permission.CRIMINAL_NETWORK,
        Permission.FINANCIAL_CRIME,
        Permission.GANG_DETECTION,
        Permission.SENSITIVE_CASE_ACCESS,
        Permission.EXPORT_REPORTS,
    },
    UserRole.INVESTIGATOR: {
        Permission.DASHBOARD,
        Permission.SEARCH_CASES,
        Permission.CHAT_ASSISTANT,
        Permission.CRIME_MAP,
        Permission.PATTERN_INTELLIGENCE,
        Permission.CRIMINAL_NETWORK,
        Permission.EXPORT_REPORTS,
    },
    UserRole.ANALYST: {
        Permission.DASHBOARD,
        Permission.SEARCH_CASES,
        Permission.CHAT_ASSISTANT,
        Permission.CRIME_MAP,
        Permission.PATTERN_INTELLIGENCE,
        Permission.EXPORT_REPORTS,
    },
    UserRole.POLICY_MAKER: {
        Permission.DASHBOARD,
        Permission.SEARCH_CASES,
        Permission.CHAT_ASSISTANT,
        Permission.CRIME_MAP,
        Permission.PATTERN_INTELLIGENCE,
    },
}


def normalize_role(role_str: str | None) -> str | None:
    """Normalize mixed case or legacy role names to canonical uppercase representation."""
    if not role_str:
        return None
    r = role_str.strip().lower().replace(" ", "_").replace("-", "_")
    if r in ("admin", "administrator"):
        return "ADMINISTRATOR"
    if r == "supervisor":
        return "SUPERVISOR"
    if r == "senior_investigator":
        return "SENIOR_INVESTIGATOR"
    if r == "investigator":
        return "INVESTIGATOR"
    if r == "analyst":
        return "ANALYST"
    if r in ("policy_maker", "policymaker"):
        return "POLICY_MAKER"
    return role_str.upper()


def check_permission(role: str | None, permission: str) -> bool:
    """Verify if a role has a specific permission with legacy fallback compatibility."""
    if not role:
        return False
    norm_role = normalize_role(role)
    perms = ROLE_PERMISSIONS.get(norm_role, set())
    
    # 1. Direct match check
    if permission in perms:
        return True
        
    # 2. Legacy fallback mapping
    legacy_map = {
        "pattern_analysis": "PATTERN_INTELLIGENCE",
        "export_data": "EXPORT_REPORTS",
        "search_cases": "SEARCH_CASES",
        "criminal_network": "CRIMINAL_NETWORK",
        "crime_map": "CRIME_MAP",
        "chat_assistant": "CHAT_ASSISTANT",
        "financial_crime": "FINANCIAL_CRIME",
        "gang_detection": "GANG_DETECTION",
        "sensitive_case_access": "SENSITIVE_CASE_ACCESS",
        "view_audit_logs": "VIEW_AUDIT_LOGS",
        "manage_users": "MANAGE_USERS",
        "system_configuration": "SYSTEM_CONFIGURATION",
    }
    
    mapped_perm = legacy_map.get(permission)
    if mapped_perm and mapped_perm in perms:
        return True
        
    # Handle direct match of mapped string
    norm_permission = permission.upper()
    if norm_permission in perms:
        return True
        
    return False
